Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs

```diff
@@ -0,0 +1,252 @@
+
+        /// <summary>Well known file which stores the default application credentials.</summary>
+        private const string WellKnownCredentialsFile = "application_default_credentials.json";
+
+        /// <summary>Environment variable which contains the Application Data settings.</summary>
+        private const string AppdataEnvironmentVariable = "APPDATA";
+
+        /// <summary>GCloud configuration directory in Windows, relative to %APPDATA%.</summary>
+        private const string CloudSDKConfigDirectory = "gcloud";
+
+        /// <summary>Help link to the application default credentials feature.</summary>
+        private const string HelpPermalink = "https://developers.google.com/accounts/docs/application-default-credentials";
+
+        private object lockObject = new object();
+
+        // These variables should only be accessed inside a synchronized block.
+        private GoogleCredential cachedCredential = null;
+        private bool checkedComputeEngine = false;
+
+        /// <summary>
+        /// Returns the Application Default Credentials
+        /// See <see cref="M:Google.Apis.Auth.OAuth2.GoogleAuth.GetApplicationDefaultCredential"/> for details.
+        /// </summary>
+        public GoogleCredential GetApplicationDefaultCredential()
+        {
+            if (cachedCredential == null)
+            {
+                lock (lockObject)
+                {
+                    if (cachedCredential == null)
+                    {
+                        cachedCredential = GetDefaultCredentialUnsynchronized();
+                    }
+                }
+            }
+            if (cachedCredential != null)
+            {
+                return cachedCredential;
+            }
+
+            throw new InvalidOperationException(String.Format("The Application Default Credentials are not available. They are available if running"
+                + " in Google Compute Engine. Otherwise, the environment variable {0} must be defined"
+                + " pointing to a file defining the credentials. See {1} for more information.",
+                CredentialEnvironmentVariable,
+                HelpPermalink));
+        }
+
+        /// <summary>Unsynchronized helper method to obtain the Application Default Credentials.</summary>
+        private GoogleCredential GetDefaultCredentialUnsynchronized()
+        {
+            string credentialPath = null;
+            
+            // 1. First try the environment variable.
+            try
+            {
+                credentialPath = GetEnvironmentVariable(CredentialEnvironmentVariable);
+                if (!String.IsNullOrWhiteSpace(credentialPath))
+                {
+                    return CreateDefaultCredentialFromFile(credentialPath);
+                }
+            }
+            catch (Exception e)
+            {
+                // Catching generic exception type because any corrupted file could manifest in different ways including 
```

Comment:
Which line from 104 to 108 can throw the exception? Let's limit the try and catch only to the throwing function.
Same below...